astropy/cosmology/funcs/optimize.py

```diff
@@ -1,31 +1,104 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """Convenience functions for `astropy.cosmology`."""
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING, Any, NotRequired, Protocol, TypedDict
 
 import numpy as np
+import numpy.typing as npt
 
 from astropy.cosmology import units as cu
 from astropy.cosmology.core import CosmologyError
 from astropy.units import Quantity
 from astropy.utils.exceptions import AstropyUserWarning
 
+if TYPE_CHECKING:
+    from collections.abc import Callable
+    from typing import TypeAlias
+
+    from scipy.optimize import OptimizeResult
+
 __all__ = ["z_at_value"]
 
 __doctest_requires__ = {"*": ["scipy"]}
 
 
+# TODO: it would be nice to upstream this to scipy and then use it here.
+class CustomMethod(Protocol):
+    """Protocol for custom :mod:`scipy.optimize.minimize_scalar` methods.
+
+    See :mod:`scipy.optimize.minimize_scalar` for details.
+    """
+
+    def __call__(
+        self, fun: Callable[..., Any], args: tuple[Any], **kwargs: Any
+    ) -> OptimizeResult: ...
+
+
+BracketSingle: TypeAlias = tuple[float, float] | tuple[float, float, float] | None
```

Comment:
If `T` is some type that `None` is not an instance of then `T` and `T | None` are different types. That makes it simple to use `T | None` to express the idea of something that should be an instance of `T` but might not have been initialized, and to use `T` to convey that a value is definitely initialized. The current definition of `BracketSingle` includes `None`, so `BracketSingle` and `BracketSingle | None` are the same type. Why wouldn't the distinction between the two be useful?

Reply:
SGTM. I'll separate it.

Reply:
Done.